File: frontend-merchant/src/types/review.ts
/**
 * Phase 5 · 评价域类型定义（商家侧视角）。
 *
 * 严格对齐 docs/API/phase-5-contracts.md §3.1 / §3.2 / §4 / §5。
 * 命名保持后端 snake_case；避免手工映射错位。
 *
 * 商家端只需要「读取本店评价 + 就地回复（创建 / 修改 / 删除）」——
 * 因此不定义评价创建 / 编辑 / 删除 / 举报相关 payload（那些属于 user 端）。
 */

// ---------- 评价（Review） --------------------------------------------------

/**
 * 单条评价对象（商家端查看）。
 * 字段来自 review 表 + 后端聚合的商品 / 用户展示信息。
 */
export interface MerchantReviewOut {
  id: number;
  order_id: number;
  order_no: string;
  order_item_id: number;
  user_id: number;
  /** 展示昵称；匿名评价后端已脱敏为「匿***名」。 */
  user_display_name: string;
  spu_id: number;
  sku_id: number;
  shop_id: number;
  /** 商品标题快照，避免联查 SPU。 */
  spu_title: string;
  /** SKU 规格摘要（如「颜色：红/尺寸：M」）。 */
  sku_specs: Record<string, string> | null;
  /** SKU 图片 object_key（可能为空）。 */
  sku_image: string | null;
  rating: number;
  content: string;
  /** MinIO object_key 数组，最多 6 张。 */
  images: string[];
  is_anonymous: boolean;
  /** 是否已被平台隐藏（商家可见但会打灰）。 */
  visible: boolean;
  hidden_reason: string | null;
  edit_count: number;
  edit_deadline_at: string;
  created_at: string;
  updated_at: string;
  /** 内联的商家回复（若已回复）。 */
  reply: MerchantReviewReplyOut | null;
}

/** 商家回复对象（评价内联）。 */
export interface MerchantReviewReplyOut {
  id: number;
  review_id: number;
  merchant_account_id: number;
  shop_id: number;
  content: string;
  created_at: string;
  updated_at: string;
}

// ---------- 评价列表查询 ----------------------------------------------------

/**
 * `GET /api/v1/merchant/reviews` 查询参数。
 * - rating 精确单值筛选（1-5）
 * - has_reply=false 只看未回复；true 只看已回复；unset 全部
 * - keyword 匹配评价内容 / 商品标题（后端实现自定）
 */
export interface MerchantReviewListQuery {
  rating?: number;
  has_reply?: boolean;
  keyword?: string;
  page?: number;
  size?: number;
}

// ---------- 评价回复 payload -------------------------------------------------

/** `POST /api/v1/merchant/reviews/{id}/reply` */
export interface CreateReplyPayload {
  /** 5-500 字。 */
  content: string;
}

/** `PATCH /api/v1/merchant/reviews/{id}/reply` */
export interface UpdateReplyPayload {
  content: string;
}

// ---------- 评价统计（商家看板） --------------------------------------------

/**
 * 评价看板 stats。
 *
 * Phase 5 契约未显式定义 merchant 端 stats 端点，
 * 但 §4.4 / §5.3 支持按 rating / has_reply 筛选；前端可以：
 *   - 直接调 `GET /merchant/reviews?has_reply=false&size=1` 获取 total（未回复数）
 *   - 直接调 `GET /merchant/reviews?rating=<n>&size=1` 获取 total（各星级数）
 *   - shop.rating_avg / rating_count 从 `GET /merchant/me`.shop 拿
 *
 * 为避免 N+1 请求，前端封装了 `useMerchantReviewStats` hook
 * 通过一次拉取 100 条最新评价在本地聚合基础指标（够 dashboard 用）。
 */
export interface MerchantReviewStats {
  total_count: number;
  unreplied_count: number;
  avg_rating: number;
  /** 差评（≤ 3 星）数量。 */
  low_rating_count: number;
  /** 各星级分布：`{5: 40, 4: 20, ...}` */
  rating_distribution: Record<number, number>;
}

File: backend/app/schemas/review.py
"""Review / review-reply / review-report schemas — Phase 5 contract §4-§5."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.review_report import ReviewReportReasonCategory, ReviewReportStatus

_MIN_CONTENT = 5
_MAX_CONTENT = 2000
_MAX_IMAGES = 6
_MIN_RATING = 1
_MAX_RATING = 5
_MIN_REPLY = 5
_MAX_REPLY = 500
_MAX_HIDE_REASON = 500
_MAX_REPORT_NOTE = 500


# ---------------------------------------------------------------------------
# Review
# ---------------------------------------------------------------------------
class ReviewItemIn(BaseModel):
    """Single order-item review payload inside the batch."""

    order_item_id: int = Field(ge=1)
    rating: int = Field(ge=_MIN_RATING, le=_MAX_RATING)
    content: str = Field(min_length=_MIN_CONTENT, max_length=_MAX_CONTENT)
    images: list[str] = Field(default_factory=list, max_length=_MAX_IMAGES)
    is_anonymous: bool = False


class ReviewCreateBatchIn(BaseModel):
    """Batch-create: one API call may cover multiple items of the same order."""

    reviews: list[ReviewItemIn] = Field(min_length=1, max_length=20)


class ReviewUpdateIn(BaseModel):
    """Edit an existing review inside the 15-day window (once)."""

    rating: int | None = Field(default=None, ge=_MIN_RATING, le=_MAX_RATING)
    content: str | None = Field(default=None, min_length=_MIN_CONTENT, max_length=_MAX_CONTENT)
    images: list[str] | None = Field(default=None, max_length=_MAX_IMAGES)

    @model_validator(mode="after")
    def _at_least_one(self) -> ReviewUpdateIn:
        if self.rating is None and self.content is None and self.images is None:
            raise ValueError("at least one of rating/content/images required")
        return self


class ReviewReplyBrief(BaseModel):
    """Merchant reply projection embedded in a Review row."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    content: str
    shop_id: int
    merchant_account_id: int
    created_at: datetime
    updated_at: datetime


class ReviewOut(BaseModel):
    """A single review projection."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    order_id: int
    order_item_id: int
    user_id: int
    user_display_name: str | None = None  # "匿***名" when anonymous
    spu_id: int
    sku_id: int
    shop_id: int
    rating: int
    content: str
    images: list[str] = Field(default_factory=list)
    is_anonymous: bool
    visible: bool
    hidden_reason: str | None = None
    edit_count: int
    edit_deadline_at: datetime
    created_at: datetime
    updated_at: datetime
    reply: ReviewReplyBrief | None = None


class ReviewRatingSummary(BaseModel):
    """Aggregation over a list of visible reviews."""

    avg: float
    count: int
    distribution: dict[int, int] = Field(default_factory=dict)


class ReviewListOut(BaseModel):
    items: list[ReviewOut]
    total: int
    page: int
    size: int
    summary: ReviewRatingSummary | None = None


class AdminReviewHideIn(BaseModel):
    hidden_reason: str = Field(min_length=5, max_length=_MAX_HIDE_REASON)


# ---------------------------------------------------------------------------
# Review reply
# ---------------------------------------------------------------------------
class ReviewReplyIn(BaseModel):
    content: str = Field(min_length=_MIN_REPLY, max_length=_MAX_REPLY)


class ReviewReplyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    review_id: int
    merchant_account_id: int
    shop_id: int
    content: str
    created_at: datetime
    updated_at: datetime


# ---------------------------------------------------------------------------
# Review report
# ---------------------------------------------------------------------------
class ReviewReportIn(BaseModel):
    reason_category: ReviewReportReasonCategory
    reason_note: str | None = Field(default=None, max_length=_MAX_REPORT_NOTE)


class ReviewReportOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    review_id: int
    reporter_user_id: int
    reason_category: ReviewReportReasonCategory
    reason_note: str | None = None
    status: ReviewReportStatus
    reviewer_admin_id: int | None = None
    review_note: str | None = None
    reviewed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class AdminReportHandleIn(BaseModel):
    review_note: str | None = Field(default=None, max_length=1000)


__all__ = [
    "AdminReportHandleIn",
    "AdminReviewHideIn",
    "ReviewCreateBatchIn",
    "ReviewItemIn",
    "ReviewListOut",
    "ReviewOut",
    "ReviewRatingSummary",
    "ReviewReplyBrief",
    "ReviewReplyIn",
    "ReviewReplyOut",
    "ReviewReportIn",
    "ReviewReportOut",
    "ReviewUpdateIn",
]
